File: template-method/ruby/compiler-x86.rb
require_relative 'compiler'

class CompilerX86 < Compiler
	def initialize()
		puts("\n\nInitialize X86 Compiler\n\n")
	end

	def preprocess(code)
		output = 'void main() { cout << 2147483647 << endl; }'
		puts("X86 preprocesor generated '#{output}'")
		return output
	end

	def compile(code)
		output = 'msg  db "2147483647", 0x0d, 0x0a, "$"'
		puts("X86 compiler generated '#{output}'")
		return output
	end

	def assemble(code)
		output = _to_binary(code)
		puts("X86 assembler generated '#{output}'")
		return output
	end

	def link(code)
		output = _to_binary("CompilerX86")+code
		puts("X86 linker generated '#{output}'")
		return output
	end

private

	def _to_binary(code)
		return code.bytes.map{|value| value.to_i.to_s(2)}.join[0,16]
	end
end

File: template-method/python3/compiler_x86.py
from compiler import *

class CompilerX86(Compiler):
	def __init__(self):
		super()
		print("\n\nInitialize X86 Compiler\n\n")

	def preprocess(self, code):
		output = 'void main() { cout << 2147483647 << endl; }'
		print("X86 preprocesor generated '{}'".format(output))
		return output

	def compile(self, code):
		output = 'msg  db "2147483647", 0x0d, 0x0a, "$"'
		print("X86 compiler generated '{}'".format(output))
		return output

	def assemble(self, code):
		output = self._to_binary(code)
		print("X86 assembler generated '{}'".format(output))
		return output

	def link(self, code):
		output = self._to_binary("CompilerX86")+code
		print("X86 linker generated '{}'".format(output))
		return output

	def _to_binary(self, code):
		code_bytes = ''
		for char in code:
			code_bytes += format(ord(char), '08b')
		return code_bytes[:16]
